src/BullOak.Repositories.EventStore/EventStoreSession.cs

```
@@ -0,0 +1,161 @@
+                    result.AddRange(currentSlice.Events);
+                } while (!currentSlice.IsEndOfStream);
+                currentVersion = (int)currentSlice.LastEventNumber;
+                return result;
+            }
+        }
+
+        protected override void Dispose(bool disposing)
+        {
+            if (disposing) { ConsiderSessionDisposed(); }
+            base.Dispose(disposing);
+        }
+
+        private void ConsiderSessionDisposed()
+        {
+            isInDisposedState = true;
+        }
+
+        /// <summary>
+        /// Saves changes to the respective stream
+        /// NOTES: Current implementation doesn't support cancellation token
+        /// </summary>
+        /// <param name="eventsToAdd"></param>
+        /// <param name="shouldSaveSnapshot"></param>
+        /// <param name="snapshot"></param>
+        /// <param name="cancellationToken"></param>
+        /// <returns></returns>
+        protected override async Task<int> SaveChanges(object[] eventsToAdd,
+            bool shouldSaveSnapshot,
+            TState snapshot,
+            CancellationToken? cancellationToken)
+        {
+            checked
+            {
+                CheckDisposedState();
+                var writeResult = await eventStoreConnection.ConditionalAppendToStreamAsync(
+                    streamName,
+                    currentVersion,
+                    eventsToAdd.Select(eventObject =>
+                    {
+                        return CreateEventData(eventObject);
+                    }))
+                    .ConfigureAwait(false);
+
+                switch (writeResult.Status)
+                {
+                    case ConditionalWriteStatus.Succeeded:
+                        break;
+                    case ConditionalWriteStatus.VersionMismatch:
+                        throw new InvalidOperationException($"Stream expected version mismatched actual version. StreamId: {streamName}");
```

This is the definition of a concurrency exception and should throw a concurrency exception rather than an `InvalidOperationException`